### ISSUE TITLE
Custom handling of push events

### ISSUE THREAD

**Issue by brianmoose** [idaholab/civet#389]
Currently on moose devel push events, we auto cancel any jobs on previous events. This can lead to the devel->master merge not happening until the end of the day.

The proposed alternative is to have a "current" event that will run to completion even if other push events happen after it. This ensures that we at least get some merges to the master branch throughout the day.

For example:
* event E1 happens. All jobs on E1 will get run, no matter what
* event E2 happens while E1 is running. Jobs on E2 won't start until all jobs on E1 are started. If all jobs on E1 are running/done then jobs on E2 can start.
* event E3 happens while E1 and E2 are running. This cancels all jobs on E2. The cancelling logic would cancel any event back to the first event that is running. Jobs on E3 won't start until all jobs on E1 are running/done.
* event E1 finishes. Now jobs on E3 become the "current" and will finish before any subsequent events.

@permcody and @aeslaughter Does this capture what you wanted?

**Comment by brianmoose** [comment]
One drawback of this method is the case where a bunch of stuff is merged in a short time frame. This seems to be a somewhat common occurrence in the morning for example.
We would get the first merge running to completion and then eventually the last merge. If they all happen in a short time frame, it might be better just to run the last one to completion.

We could combine the above strategy with a heuristic for doing the canceling of events. Maybe using total event job running time?

For example, if E2 above happens 5 minutes after E1, the total job running time of E1 will still be quite small even if all the jobs got picked up immediately. So we would cancel E1 and start on E2 (E2 would be the "current" event and would potentially be run to completion).

If E3 happens a hour after E2, the total job running time of E2 would get checked. If hardly any jobs have been picked up (due to there being a bunch of PRs using the build boxes) then the total job running time would be small and we would cancel E2 and start jobs on E3. If, however, E2 has been running a while, then E2 would remain the "current" event and finish to completion.

Basically we would setting a "point of no return" for an event. After a threshold of total job running time has been exceeded then that event would run to completion. If it hasn't been exceeded, then it can still be auto cancelled.

Thoughts?

**Comment by permcody** [comment]
I'm not a huge fan of any heuristic here or time-based strategy because it will constantly discourage devs to merge. We'll hear people say, "well, we merge it after lunch because CIVET has only been working on it for 30 mins and will cancel... or it's almost done so we need to let CIVET catch up".

I see your point that the first merge of the morning will just go through and the second through fifth ones won't even be tested for a few hours. I still think it's a better strategy because it won't discourage merges of any kind at any time. 

One way we could get around the drawback you mentioned is allow a human to cancel any job in the list, or perhaps even a new button that just manually cancels the "current" merge event as we are calling it. If we had that, CIVET should recognize that it's not possible to complete that event and move to the next strategy (designate a new "current").
